pkg/imgpkg/image/keychain.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+package image
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	regauthn "github.com/google/go-containerregistry/pkg/authn"
+	regname "github.com/google/go-containerregistry/pkg/name"
+)
+
+// constants copied from https://github.com/vmware-tanzu/carvel-imgpkg/blob/c8b1bc196e5f1af82e6df8c36c290940169aa896/vendor/github.com/docker/docker-credential-helpers/credentials/error.go#L4-L11
+const (
+	// ErrCredentialsNotFound standardizes the not found error, so every helper returns
+	// the same message and docker can handle it properly.
+	errCredentialsNotFoundMessage = "credentials not found in native keychain"
+	// ErrCredentialsMissingServerURL and ErrCredentialsMissingUsername standardize
+	// invalid credentials or credentials management operations
+	errCredentialsMissingServerURLMessage = "no credentials server URL"
+	errCredentialsMissingUsernameMessage  = "no credentials username"
+
+	globalPrefix = "IMGPKG_REGISTRY"
+)
+
+func Keychain(username string, password string, token string, anon bool) regauthn.Keychain {
```

Comment:
I like the second suggestion because this function can already filter out the environment variables that we care about so that we do not have to go through all of them.